### PR TITLE
Feature/74-75-76-77-128-roles-authorisations

### PR DESCRIPTION
**Pull request by JuanluRM**
This pull request will update the permissions for non admin roles, the permissions are based on requirements and are described as follow:

-Educators can CRUD StudentsEvaluation and QuarterMarks. Can get any lesson but only modify the lessons they are educators of and can't modify the educators field. Can get and modify LessonAtendance and EventLesson but only the educators field. Can get Students.
-Volunteers can get and modify LessonAttendance. Can get and modify Events and EventLessons but only the volunteers field. 
-Families can CRUD Students. Can get QuarterMarks, StudentEvaluations and Lessons. Can get and modify Events and EventLessons but only the attendees field.
-Partners can get Events, LessonEvents and Meetings. They can get and post donations and punctualDonations.

Feel free to notify me for any permissions I've got wrong or missed from requirements.

### CODE REVIEW

**Review by auroranavas — COMMENTED** [review]
I haven't checked the code yet, but in your PR description I noticed a few misconceptions. I took the text from the PR and added in **bold** the missing details. Maybe @pabpercab1 can double-check this:
- Educators can CRUD StudentsEvaluation and **EvaluationType** **(QuarterMarks are the typical "boletines" that kids are given by their real schools every quarter. Their families are the ones that can CRUD them. Educators can GET the QuarterMarks)**. Can get any lesson but only modify the lessons they are educators of and can't modify the educators field. Can get LessonAtendance **(LessonAttendance represents a volunteer's attendance to a lesson. They can attend certain lessons to help)**. They can **GET** EventLesson but only **PUT the educators or attendees field** **(If modifying only one attribute is complicated, you can remove the PUT method from EventLesson)**. Can get Students. **Can GET Lesson and CenterAuthorization.**
- Volunteers can **CRU** LessonAttendance. Can get and modify Events and EventLessons but only the volunteers field.
-Families can CRUD Students. Can **CRUD QuarterMarks**, **GET** StudentEvaluations and **GET** Lessons. Can get and modify Events ~and EventLessons~ **(for LessonEvent, a Student can go if they have been authorized via CentreAuthorization, so the Admin or the Educator would be who adds them to the Event)** but only the attendees field. 
- Partners can **get Events and LessonEvents**.  **Can GET and PUT Meetings but only the attendees field**. They can get and post donations.

If you have any questions, feel free to ask! By the way, try to fix the Codacy error 👍.

**Comment by JuanluRM** [comment]
Everything was changed acording to your feedback @auroranavas, and should be working now. Aditionaly has been added a restriction that families can only create or update students that belongs to their family ID.

**Review by Ahydul — CHANGES_REQUESTED** [review]
I get an error 'No module named 'environ''. Maybe you need to update the requirements.txt file if you added a new requirement.

**Comment by JuanluRM** [comment]
It was caused by an error from another user on develop, should be fixed now.

**Review by auroranavas — CHANGES_REQUESTED** [review]
You did a really good job. However there are a few things that could be done to make your implementation a bit better:

**General permissions**
- Punctual Donations should only be deleted by the admin.
- HomeDocument should only be manipulated by the admin.
- Suggestions can only be deleted by the admin.

**Volunteer permissions**
- Event should only be PUT by volunteers whose 'state' is ACCEPTED. Right now, any volunteer can perform a PUT operation. 

**Educator permissions**
- `http://127.0.0.1:8000/api/evaluation-type/1/` gives this error `"detail": "You do not have permission to perform this action."`.

**Partner permissions**
- Partners should not be able to manipulate LessonEvent.

**Family permissions**
- CenterExit should be CRUDed by a family since they are the ones that upload or not authorizations.

**EducationCenter permissions**
- Basically, they should have the same permissions as a Family. 

**General user permissions**
- Educator, Partner, Family, Volunteer, EducationCenter and User should only be manipulated either by the admin or themselves. However, I think you should change this in your admin PR since it would be difficult to test this implementation without permissions to change stuff. So I would first put this PR in develop and then the admin one.

**Comment by JuanluRM** [comment]
Everything was changed according to feedback except the admin permissions because it will be changed on the admin permissions branch.

**Review by auroranavas — CHANGES_REQUESTED** [review]
Thank you for taking into account my feedback. But I did not see any of the **General permissions** reflected and I was wrong about **Family permissions** and **EducationCenter permissions**:

**General permissions**
You did not change any of these permissions. Please change them by looking at the Request for Change I wrote yesterday.

**Partner permissions**
- Partners shouldn't be able to even GET LessonEvent, as they are class-specific events and, in theory, can't attend.

**Family permissions**
- I was wrong about this, I apologize. Actually, when a LessonEvent is created, the CenterAuthorization instances for each student in the lesson should be automatically created (frontend's job), so I guess they should be CRUDed by the admin and PUT by the Family.
 
**EducationCenter permissions**
- Same change as Family.

**Comment by JuanluRM** [comment]
Everything was changed as we accorded

**Review by auroranavas — APPROVED** [review]
Thank you for bearing with my feedback. You did a really good job and you put effort into understanding the problem. Congrats! 😊

**Review by Ahydul — APPROVED** [review]
Everything is correct

**Review by pabpercab1 — APPROVED** [review]
Well done creating the authorisations. You have also considered the feedback of other colleagues as well. Nice one!
